rfd/0-job-worker.md

```diff
@@ -0,0 +1,240 @@
+- complete: the job completed successfully on its own
+- stopped: the job was manually stopped, this takes precedence over complete
+
+##### Job Output Storage
+
+Output for each job will be maintained, in memory, as long as the worker implementation exists. This poses a potential memory leak, but for the purposes of this challenge, is an acceptable solution. Later designs would likely need to write job output to disk or possibly even remote storage.
+
+##### Job Output Streaming
+
+The worker library streams job output through an `io.ReadCloser`. The server will indicate the end of the stream, if the job has completed, with the `io.EOF` error. The client may, at any time, close the stream to indicate that it is disconnecting. The client must always close the stream when it no longer needs it to prevent memory leaks.
+
+Internally, the library will maintain a buffer containing the job output. The job itself will be configured to use this buffer for `stdout` and `stderr` and the `Write()` method will be goroutine safe. When a client calls `JobOutput()` the library will create a new `io.Reader` that independently, and with goroutine safety, reads through to the end of the output buffer. If the job has already completed at this time, `io.EOF` will be returned. If not, subsequent calls to `Read()` will block until there is either more output to return or the job ends in which case `io.EOF` is returned. When the job calls `Write()` it will signal connected readers that new data has been written to the output buffer. The readers will in turn be able to unblock and return their `Read()` calls to connected clients at that time.
+
+##### Stopping Jobs
+
+StopJob is an asynchronous request that takes an optional timeout. If the timeout is greater than 0, it will cause the library to first issue a `SIGINT` and wait up to `timeout` for the job to complete. If the job doesn't complete after `timeout`, or timeout was `0`, then the job is terminated with `SIGKILL`. The library returns a channel that will be closed when the job completes.
+
+##### Identifiers
+
+JobID is an opaque identifier that is internally mapped to job in the library. Since pid is not guaranteed to be unique, nothing internal to the library will be keyed off of the pid. The library must still track the pid though in order to signal the process to stop, for example. The `go.jetify.com/typeid` library is used to generate job ids.
+
+UserID can be any unique value as far as the library is concerned. See the section on Authorization below for more information.
+
+#### gRPC API
+
+The gRPC API follows the worker library fairly closely. The only significant difference is that it does not require the user id in the messages because user id is the client certificate's subject, not explicitly provided. Additionally, `StopJob()` in the library returns a channel that closes when the job completes. The gRPC API does not provide a facility for notifying the client when the job completes.
```

Comment:
> the client certificate's subject

What part of the Subject?

Reply:
i was planning on just using `(*x509.Certificate).Subject.String()` as it looks to be repeatable between certs. I'll clarify in the doc.

Reply:
updated